### DISCUSSION TITLE
About match algoritm

### DISCUSSION THREAD

**Discussion by APerricone** [cssbattle/forum#269] [Bugs reports]
Hello, I am having some throuble with the scoring system, as pointed in https://github.com/cssbattle/forum/discussions/248#discussioncomment-2618571 looks like the algorith that evaluates the result is buggy.
Using cssBattle previewer 2 I was able to download a couple of result that look like thet are not eveluated correctly, the first one with "diff" comparing gets this
![image](https://user-images.githubusercontent.com/22457168/172988348-b540e5a7-2ece-438a-a0f5-02c68681febf.png)
and obtains 99.9%
the second one:
![image](https://user-images.githubusercontent.com/22457168/172988428-902afe2c-c49d-4694-adf3-5c2656a1e401.png)
and obtains 100.0%.
I wrote a my version of the comparing algorith
```js
                    canvas.width = imgOri.width; //400
                    canvas.height = imgOri.height; //300
                    let c2d = canvas.getContext('2d')
                    c2d.drawImage(imgOri, 0, 0, imgOri.width, imgOri.height);
                    let oriID = c2d.getImageData(0,0,imgOri.width,imgOri.height).data
                    c2d.drawImage(imgNew, 0, 0, imgNew.width, imgNew.height);
                    let newID = c2d.getImageData(0,0,imgNew.width,imgNew.height).data
                    final = new ImageData(imgNew.width, imgNew.height)
                    let diff = 0;
                    for(let i=0,y=0;y<canvas.height;++y) for(let x=0;x<canvas.width;++x) for(let b=0;b<4;++b,++i) {
                        if(b==3) final.data[i]=255; else {
                            let v = Math.abs(oriID[i]-newID[i]);
                            diff += v;
                            final.data[i]=v;
                            if(b==3) final.data[i]=255;
                        }
                    }
                    c2d.putImageData(final,0,0)
                    let percent = 100-(diff*100 / (canvas.width*canvas.height*4*255));
```
and the results are 99.97512826797386% and 99.92589869281046%, that are values I expect.

I put my tests on http://amicoperry.altervista.org/others/CSSBattleTest/index.html
please, can the mainteiners verify?

**Comment by Beowolve** [comment]
I think the checking works differently.
There seams to be a max distance for each pixel that is individuallly checked first. A global distance is maybe there as well, but I am not so sure about that.
Thats why I didn't add a match % yet, I didn't want to give a wrong indication in the previewer. If the matching can be replicated correctly then this would make much more sense.

Anyway, nice comparisson 👍

**Comment by APerricone** [comment]
I think it is possible change the color's distance algorithm, using gamma correction or another color space, but even with those modification the first one is closer than second. For what I see the game is create a random image without coherent criteria and hope for the result, and it is not the game I want play

**Comment by APerricone** [comment]
on 114:
```html
<a a><a b><a c><a d>
<style>a{position:fixed;border:solid;border-width:75 200}
[a]{border-color:#000 #172D2C #E9AF53 ;top:0;right:200}
[b]{border-color:#172D2C #E9AF53;top:0;left:200}
[c]{border-color:#E9AF53 #E9AF53 #172D2C;bottom:0;right:200}
[d]{border-color:#172D2C #172D2C #000;bottom:0;left:200}
```
can someone explain where is the 99.9% match?

**Reply by rafehimself** [reply]
it's bugged,
I remember in #79 lower arc was also bugged, what got 100% wasn't exactly what was matching in the viewer
